File: src/test/java/bop/c/MemoryTest.java
package bop.c;

import bop.bench.Bench;
import java.lang.foreign.AddressLayout;
import java.util.stream.IntStream;
import jdk.internal.foreign.SegmentFactories;
import org.junit.jupiter.api.Test;

public class MemoryTest {
  @Test
  public void benchFFMOverhead() throws Throwable {
    Bench.printHeader();
    IntStream.of(1, 2, 4, 8, 16).forEach(threads -> {
      try {
        Bench.threaded("c stub", threads, 10, 10000000, (threadId, cycle, iteration) -> {
          Memory.hello();
        });
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    });
    Bench.printFooter();
  }

  @Test
  public void alloc() {
    final var address = Memory.alloc(12);
    System.out.println(address);
    Memory.dealloc(address);

    final var segment = SegmentFactories.makeNativeSegmentUnchecked(Memory.alloc(64L), 64L);
    //    final var segment = MemorySegment.ofAddress(Memory.alloc(64L));
    segment.set(AddressLayout.JAVA_LONG, 0, address);

    final var addr = segment.get(AddressLayout.JAVA_LONG, 0);
    System.out.println(addr);
  }

  static final ThreadLocal<byte[]> BYTE_LOCAL = new ThreadLocal<>();

  @Test
  public void benchHeapAccess() throws Throwable {
    Bench.printHeader();
    IntStream.of(1, 2, 4, 8).forEach(threads -> {
      try {
        Bench.threaded("bop_heap_access", threads, 25, 10000000, (threadId, cycle, iteration) -> {
          byte[] d = BYTE_LOCAL.get();
          if (d == null) {
            d = new byte[16];
            BYTE_LOCAL.set(d);
          }
          Memory.heapAccess(d);
        });
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    });
    Bench.printFooter();
  }

  @Test
  public void benchAllocDealloc() throws Throwable {
    Bench.printHeader();
    IntStream.of(1, 2, 4, 8, 16).forEach(threads -> {
      try {
        Bench.threaded(
            "bop_alloc/bop_dealloc", threads, 10, 10000000, (threadId, cycle, iteration) -> {
              Memory.dealloc(Memory.alloc(16L));
            });
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    });
    Bench.printFooter();
  }

  @Test
  public void benchZallocDealloc() throws Throwable {
    Bench.printHeader();
    IntStream.of(1, 2, 4, 8, 16).forEach(threads -> {
      try {
        Bench.threaded(
            "bop_alloc/bop_dealloc", threads, 10, 10000000, (threadId, cycle, iteration) -> {
              Memory.dealloc(Memory.zalloc(16L));
            });
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    });
    Bench.printFooter();
  }
}

File: jvm/core/src/test/java/bop/c/MemoryTest.kt
package bop.c

import bop.bench.Bench
import bop.bench.Operation
import org.junit.jupiter.api.Test
import java.util.function.IntConsumer
import java.util.stream.IntStream

class MemoryTest {
   @Test
   fun benchFFMOverhead() {
      Bench.printHeader()
      IntStream.of(1, 2, 4, 8, 16).forEach(IntConsumer { threads: Int ->
         try {
            Bench.threaded(
               "c stub",
               threads,
               10,
               10000000,
               Operation { threadId: Int, cycle: Int, iteration: Int ->
                  Memory.hello()
               })
         } catch (e: InterruptedException) {
            throw RuntimeException(e)
         }
      })
      Bench.printFooter()
   }

   @Test
   fun alloc() {
      val address = Memory.alloc(12)
      println(address)
      Memory.dealloc(address)
   }

   @Test
   fun benchHeapAccess() {
      Bench.printHeader()
      IntStream.of(1, 2, 4, 8).forEach(IntConsumer { threads: Int ->
         try {
            Bench.threaded(
               "bop_heap_access",
               threads,
               25,
               10000000,
               Operation { threadId: Int, cycle: Int, iteration: Int ->
                  var d: ByteArray? = BYTE_LOCAL.get()
                  if (d == null) {
                     d = ByteArray(16)
                     BYTE_LOCAL.set(d)
                  }
                  Memory.heapAccess(d)
               })
         } catch (e: InterruptedException) {
            throw RuntimeException(e)
         }
      })
      Bench.printFooter()
   }

   @Test
   fun benchAllocDealloc() {
      Bench.printHeader()
      IntStream.of(1, 2, 4, 8, 16).forEach(IntConsumer { threads: Int ->
         try {
            Bench.threaded(
               "bop_alloc/bop_dealloc",
               threads,
               10,
               10000000,
               Operation { threadId: Int, cycle: Int, iteration: Int ->
                  Memory.dealloc(
                     Memory.alloc(16L)
                  )
               })
         } catch (e: InterruptedException) {
            throw RuntimeException(e)
         }
      })
      Bench.printFooter()
   }

   @Test
   fun benchZallocDealloc() {
      Bench.printHeader()
      IntStream.of(1, 2, 4, 8, 16).forEach(IntConsumer { threads: Int ->
         try {
            Bench.threaded(
               "bop_alloc/bop_dealloc",
               threads,
               10,
               10000000,
               Operation { threadId: Int, cycle: Int, iteration: Int ->
                  Memory.dealloc(
                     Memory.zalloc(16L)
                  )
               })
         } catch (e: InterruptedException) {
            throw RuntimeException(e)
         }
      })
      Bench.printFooter()
   }

   companion object {
      val BYTE_LOCAL: ThreadLocal<ByteArray?> = ThreadLocal<ByteArray?>()
   }
}
